### PR TITLE
Add `and_invoke` for sequential mixed (return/raise) responses.

### PR DESCRIPTION
**Pull request by askreet**
In tests I occasionally encounter things like this:

```ruby
called = false
allow(mock_api).to receive(:put_something) do
  # raise first call, second call succeeds
  unless called
    called = true
    raise ApiError, "Some Failure!"
  end
  :some_useful_value
end
```

It's commonly suggested to do similar things on [StackOverflow](https://stackoverflow.com/questions/14206477/rspec-retry-throw-exception-and-then-return-value), for example. It seemed a common enough problem to try to generalize a solution that is a bit more readable. This patch allows the following alternative:

```ruby
allow(mock_api).to receive(:put_something).and_invoke(-> { raise ApiError, "Some Failure!" },
                                                      -> { :some_useful_value })
```

`and_invoke` behaves like `and_return` with regards to sequential calls, matching expected call counts, and repeating the last call indefinitely in cases where `allow` is used vs. `expect` with `exactly(n).times`.

I alternatively considered some more magical approaches like:

```ruby
allow(mock_api).to receive(:put_something).and_return_or_raise(ApiError.new("Some Failure!"), true)
```

```ruby
allow(mock_api).to receive(:put_something).chain_response
                                          .and_raise(RuntimeError, "Some Failure!")
                                          .and_return(true)
```

Ultimately, I thought Procs were more "native", if slightly ugly.

### CODE REVIEW

**Comment by askreet** [comment]
Ah, did not realize `rspec-mocks` kept compatibility with Ruby 1.8.7 -- impressive! Will fix the syntax to be compatible. (I assume this is just a "stabby lambda" issue.)

**Comment by pirj** [comment]
How do you feel about accumulating configured responses?
Now, e.g. `and_return` accepts several arguments, but the following:
```ruby
allow(foo)
  .to receive(:bar)
  .and_return(1)
  .and_return(2)
```
won't work.

In the same way, it could be possible to use `and_raise`.

It's very similar to what you suggest with `chain_response`.

Side note: it's kind of confusing that it is possible to configure several responses at the moment, but only one, the last defined, actually comes into play.

**Comment by askreet** [comment]
I'm surprised this works, too, @pirj. I'm a bit worried about changing it because in all liklihood there are people who have done something like:

```ruby
def default_fixture
  instance_double(Foo).tap |dbl|
    expect(dbl).to receive(:color).and_return(:blue)
  end
end

it "bar" do
  expect(baz(default_fixture)).to eq(4)
end

it "bars slightly differently under certain circumstances" do
  fixture = default_fixture.and_return(:aqua)

  expect(baz(fixture)).to eq(5)
end
```

However, your example raises a neat idea that would require a bit more careful architecture change:

```ruby
allow(foo)
  .to receive(:bar)
  .and_return(1) # <== sets terminal action to AndReturnImplementation.new([1])
  .and_then_return(2) # <== replaces terminal action with ChainedTerminalAction.new([ existing_action, AndReturnImplementation([2]) ])
  .and_then_raise(RuntimeError, "failure!") # <== calls ChainedTerminalAction.append(AndRaiseImplementation.new(...))
```

The trick here would be the earlier chained implementations need to be extends to support _not_ repeating the last element in their list, or track if they have been 'consumed' or not so that the chained action can move onto the next item. Likewise, if we want to keep existing behavior, we'd have to repeat the last action of the last chain element. Alternatively we could say that once you opt into this chaining construct, you must build an explicit chain of response elements.

Overall, I quite like this approach, but I'll hold off on making the change until I hear from more maintainers about the direction.

Note that my original `#chained_responses` implementation is actually easier to pull off because it can set the terminal implementation to a new ChainedResponses instance and then return that, allowing all further calls to go to the new object (although that would mean you couldn't use more RSpec DSL at that point).

**Comment by pirj** [comment]
I believe that chaining configured response qualifiers would imply those responses being called in order.

It needs a second thought on how countable qualifiers should work though.
```ruby
receive(:foo)
  .and_return(1).twice
  .and_return(2).once # return 1, 1, then 2 
```
vs

```ruby
receive(:foo)
  .twice
  .and_return(1)
  .and_return(2) # ok
```

```ruby
receive(:foo)
  .thrice
  .and_return(1)
  .and_return(2) # the number of configured responses doesn't add up. Should we return `nil` on the third call?
```

```ruby
receive(:foo)
  .twice
  .and_return(1)
  .and_return(2)
  .and_return(3) # Error! The number of configured responses exceeds the expected number of calls
```

Semi-related https://github.com/rspec/rspec-expectations/issues/1252

**Comment by askreet** [comment]
I don't know, are you saying that in the case where we've qualified the counts with _expectations_ that the responses should switch from being overridden to appending in a chained fashion? The inconsistency with the behavior when you call `and_return` twice (override) is pretty surprising. The nice thing about introducing a new method (`and_then...` or `chained_responses`) is that it's clear you're altering the default behavior.

Even in qualified cases where expectation counts have been specified, people may still be relying on the override behavior (like my example from above). It seems backwards compatibility is a reasonable top priority for this gem.

File: lib/rspec/mocks/message_expectation.rb
Comment (JonRowe) [inline lib/rspec/mocks/message_expectation.rb:98]:
```suggestion
      #   allow(api).to receive(:get_foo).and_invoke(-> { raise ApiTimeout })
```

File: lib/rspec/mocks/message_expectation.rb
Comment (JonRowe) [inline lib/rspec/mocks/message_expectation.rb:102]:
```suggestion
      #   allow(api).to receive(:get_foo).and_invoke(-> { raise ApiTimeout }, -> { raise ApiTimeout }, -> { :a_foo })
```

File: lib/rspec/mocks/message_expectation.rb
Comment (JonRowe) [inline lib/rspec/mocks/message_expectation.rb:736]: This has a bug if there are nils or an empty argument splat is used.

File: lib/rspec/mocks/message_expectation.rb
Comment (JonRowe) [inline lib/rspec/mocks/message_expectation.rb:731]: The arguments should probably be passed to the procs similar to how blocks are called...

File: spec/rspec/mocks/and_invoke_spec.rb
Comment (JonRowe) [inline spec/rspec/mocks/and_invoke_spec.rb:16]:
```suggestion
          expect { allow(obj).to receive(:foo).and_invoke }.to raise_error(ArgumentError)
```

**Comment by JonRowe** [comment]
Initially I was going to suggest that as `and_return` already allows multiple static responses[1], it could be re-written to allow this scenario too, as having another name for configuring return responses seems sub-ideal, but I think I'm convinced this specific scenario needs its own name, to distinguish between a list of callables and others.

[1](https://relishapp.com/rspec/rspec-mocks/v/3-10/docs/configuring-responses/returning-a-value#specify-different-return-values-for-multiple-calls)

File: lib/rspec/mocks/message_expectation.rb
Comment (askreet) [reply lib/rspec/mocks/message_expectation.rb:731]: The original intent was to chain returns and rescues, but I don't see why this can't share the same block semantics, will fix.

File: lib/rspec/mocks/message_expectation.rb
Comment (askreet) [reply lib/rspec/mocks/message_expectation.rb:736]: Good catch -- I'll validate the inputs immediately and raise some error (will try to find prior art).

File: lib/rspec/mocks/message_expectation.rb
Comment (JonRowe) [reply lib/rspec/mocks/message_expectation.rb:731]: Chain returns is easy with `and_return(1,2,3,4)` etc, so this is a custom implementation so should respect the same rules, e.g. `and_return { |args| expect(args).to eq [] }` is valid and used to validate arguments, so its surprising if this doesn't also have access to args, as its essentially the "chain" form of `and_return { ... }`

Comment (askreet) [reply lib/rspec/mocks/message_expectation.rb]: Agreed, implemented in latest commit. :+1:

File: lib/rspec/mocks/message_expectation.rb
Comment (askreet) [reply lib/rspec/mocks/message_expectation.rb]: Should be fixed in the last couple commits. :+1:

**Comment by benoittgt** [comment]
Thanks for this proposal. I love the idea. Last version of code looks good to me. Thanks for the clarity of the PR.

**Comment by askreet** [comment]
@benoittgt Great, glad to hear it. I'm good with where this is at whenever we have quorum to merge. Thanks!

File: features/configuring_responses/mixed_responses.feature
Comment (pirj) [inline features/configuring_responses/mixed_responses.feature:16]: The feature description says "if the message is received additional times", maybe add an additional line repeating this one to prove/indicate that?

File: lib/rspec/mocks/message_expectation.rb
Comment (pirj) [inline lib/rspec/mocks/message_expectation.rb:119]: 👍 I clearly understand that you enforce passing at least a single argument and rely on Ruby.

The mutating `unshift` slightly concerned me, in the following scenario:
```ruby
procs = [-> { 1 }, -> { raise }]
allow(foo).to receive(:bar).and_invoke(*procs)
allow(foo).to receive(:baz).and_invoke(*procs)
```
but on a synthetic test, it seems that it has no effect on this user-defined `procs`, so all good.

File: lib/rspec/mocks/message_expectation.rb
Comment (pirj) [inline lib/rspec/mocks/message_expectation.rb:736]: We could check for `@procs_to_invoke.one?`, it's [there since 1.8.7](https://ruby-doc.org/core-1.8.7/Enumerable.html#method-i-one-3F).

File: lib/rspec/mocks/message_expectation.rb
Comment (pirj) [inline lib/rspec/mocks/message_expectation.rb:95]: We seem to pass message args to that proc, does it make sense to reflect this in the docs?

**Comment by JonRowe** [comment]
Thanks!

File: spec/rspec/mocks/and_invoke_spec.rb
Comment (pirj) [inline spec/rspec/mocks/and_invoke_spec.rb:25]: Personally, I would prefer not to DRY this.

File: spec/rspec/mocks/and_invoke_spec.rb
Comment (pirj) [inline spec/rspec/mocks/and_invoke_spec.rb:36]: Probably `allow`, not `expect`?

File: spec/rspec/mocks/and_invoke_spec.rb
Comment (pirj) [inline spec/rspec/mocks/and_invoke_spec.rb:42]: Do you mind adding an example showing that the block is also passed to that proc for completeness?

**Review by pirj — APPROVED** [review]
Oh, I'm late for the party 🐌 

It looks amazing, thanks for the contribution!

File: lib/rspec/mocks/message_expectation.rb
Comment (JonRowe) [reply lib/rspec/mocks/message_expectation.rb:119]: Its not a directly passed in array, as its essentially the same as `and_invoke(-> { 1 }, -> { raise })` its also identical to our `and_return` implementation (I squashed a different quibble I had due to that).

**Pull request merged**
